### PR TITLE
Testcontainers support

### PR DESCRIPTION
**Pull request by spencergibb**
Only for one test.

Need to setup so that only one container is run per module at the most.

Fixes gh-623

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -28,6 +28,7 @@
 		<spring-cloud-deployer.version>1.0.3.RELEASE</spring-cloud-deployer.version>
 		<spring-cloud-openfeign.version>2.2.3.BUILD-SNAPSHOT</spring-cloud-openfeign.version>
 		<spring-cloud-stream.version>Horsham.SR3</spring-cloud-stream.version>
+		<testcontainers.version>1.13.0</testcontainers.version>
```
Comment (bsideup) [inline pom.xml:31]:
```suggestion
		<testcontainers.version>1.14.1</testcontainers.version>
```

File: spring-cloud-consul-core/src/test/java/org/springframework/cloud/consul/test/ConsulTestcontainers.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	@Override
+	public void initialize(ConfigurableApplicationContext context) {
+		start();
+
+		MutablePropertySources sources = context.getEnvironment().getPropertySources();
+
+		if (!sources.contains("consulTestcontainer")) {
+			Integer mappedPort = consul.getMappedPort(8500);
+			HashMap<String, Object> map = new HashMap<>();
+			map.put("spring.cloud.consul.port", String.valueOf(mappedPort));
```
Comment (bsideup) [inline spring-cloud-consul-core/src/test/java/org/springframework/cloud/consul/test/ConsulTestcontainers.java:51]: it is important to also set the host, see https://www.testcontainers.org/features/networking/#getting-the-container-ip-address

File: spring-cloud-consul-core/src/test/java/org/springframework/cloud/consul/test/ConsulTestcontainers.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+			sources.addFirst(new MapPropertySource("consulTestcontainer", map));
+		}
+	}
+
+	public static void start() {
+		if (!consul.isRunning()) {
+			consul.start();
+			consul.followOutput(new Slf4jLogConsumer(logger).withSeparateOutputStreams());
```
Comment (bsideup) [inline spring-cloud-consul-core/src/test/java/org/springframework/cloud/consul/test/ConsulTestcontainers.java:60]: Consider using `consul.withLogConsumer(new Slf4jLogConsumer(logger).withSeparateOutputStreams())` during the container creation

File: spring-cloud-consul-discovery/src/test/java/org/springframework/cloud/consul/discovery/ConsulServerListTests.java [hunk 4]
```diff
@@ -39,7 +40,9 @@
 
 	@Test
 	public void tagsWork() {
-		ConsulClient consul = new ConsulClient();
+		ConsulTestcontainers.start();
+		ConsulClient consul = new ConsulClient("localhost",
```
Comment (bsideup) [inline spring-cloud-consul-discovery/src/test/java/org/springframework/cloud/consul/discovery/ConsulServerListTests.java:44]:
```suggestion
		ConsulClient consul = new ConsulClient(ConsulTestcontainers.getHost(),
```

File: spring-cloud-consul-discovery/src/test/java/org/springframework/cloud/consul/discovery/configclient/DiscoveryClientConfigServiceAutoConfigurationTests.java [hunk 5]
```diff
@@ -63,8 +64,12 @@ public void close() {
 
 	@Test
 	public void onWhenRequested() throws Exception {
+		ConsulTestcontainers.start();
+		Integer port = ConsulTestcontainers.getPort();
 		setup("server.port=0", "spring.cloud.config.discovery.enabled=true",
+				"spring.cloud.consul.port=" + port,
```
Comment (bsideup) [inline spring-cloud-consul-discovery/src/test/java/org/springframework/cloud/consul/discovery/configclient/DiscoveryClientConfigServiceAutoConfigurationTests.java:70]: don't forget to set the host

File: spring-cloud-consul-core/src/test/java/org/springframework/cloud/consul/test/ConsulTestcontainers.java (hunk 2 above)
Comment (spencergibb) [reply spring-cloud-consul-core/src/test/java/org/springframework/cloud/consul/test/ConsulTestcontainers.java]: Done

File: spring-cloud-consul-core/src/test/java/org/springframework/cloud/consul/test/ConsulTestcontainers.java (hunk 3 above)
Comment (spencergibb) [reply spring-cloud-consul-core/src/test/java/org/springframework/cloud/consul/test/ConsulTestcontainers.java]: Done

File: spring-cloud-consul-discovery/src/test/java/org/springframework/cloud/consul/discovery/configclient/DiscoveryClientConfigServiceAutoConfigurationTests.java (hunk 5 above)
Comment (spencergibb) [reply spring-cloud-consul-discovery/src/test/java/org/springframework/cloud/consul/discovery/configclient/DiscoveryClientConfigServiceAutoConfigurationTests.java]: done

File: spring-cloud-consul-core/src/test/java/org/springframework/cloud/consul/test/ConsulTestcontainers.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+			map.put(ConsulProperties.PREFIX + ".host", consul.getContainerIpAddress());
+
+			sources.addFirst(new MapPropertySource("consulTestcontainer", map));
+		}
+	}
+
+	public static void start() {
+		if (!consul.isRunning()) {
+			consul.start();
+			consul.withLogConsumer(
```
Comment (bsideup) [inline spring-cloud-consul-core/src/test/java/org/springframework/cloud/consul/test/ConsulTestcontainers.java:62]: `withLogConsumer` _after_ `start()` is a no-op, it is a configuration-time setting (you can put it next to `waitingFor`, `withExposedPorts` and others)

Comment (spencergibb) [reply spring-cloud-consul-core/src/test/java/org/springframework/cloud/consul/test/ConsulTestcontainers.java:62]: ah, thanks

Comment (spencergibb) [reply spring-cloud-consul-core/src/test/java/org/springframework/cloud/consul/test/ConsulTestcontainers.java]: done, hopefully

File: spring-cloud-consul-core/src/test/java/org/springframework/cloud/consul/test/ConsulTestcontainers.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+			HashMap<String, Object> map = new HashMap<>();
+			map.put(ConsulProperties.PREFIX + ".port", String.valueOf(mappedPort));
+			map.put(ConsulProperties.PREFIX + ".host", consul.getContainerIpAddress());
+
+			sources.addFirst(new MapPropertySource("consulTestcontainer", map));
+		}
+	}
+
+	public static void start() {
+		if (!consul.isRunning()) {
```
Comment (bsideup) [inline spring-cloud-consul-core/src/test/java/org/springframework/cloud/consul/test/ConsulTestcontainers.java:61]: nit: `start()` is idempotent, so it is safe to call it without checking the `isRunning()` flag

Comment (spencergibb) [reply spring-cloud-consul-core/src/test/java/org/springframework/cloud/consul/test/ConsulTestcontainers.java]: done

**Comment by bsideup** [comment]
See https://www.testcontainers.org/supported_docker_environment/continuous_integration/circle_ci/ on how to configure CircleCI so that it gives you access to Docker

**Comment by spencergibb** [comment]
Remaining items

- [ ]  Migrate core module
- [ ]  Migrate config module
- [ ]  Circle integration (see [testcontainers docs](https://www.testcontainers.org/supported_docker_environment/continuous_integration/circle_ci/) and [circle docs](https://circleci.com/docs/2.0/building-docker-images/)
- [ ] Jenkins integration
- [ ] Remove consul install and run for ci

**Pull request merged**